pkg/agentprofile/agent_profile.go

```diff
@@ -28,64 +29,74 @@ const (
-			continue
-		}
-
-		for _, node := range nodes {
-			matchesNode, err := profileMatchesNode(&profile, node.Labels)
-			if err != nil {
-				return nil, nil, err
-			}
+// ProfileToApply validates a profile spec and returns a map that maps each
+// node name to the profile that should be applied to it.
+func ProfileToApply(logger logr.Logger, profile *datadoghqv1alpha1.DatadogAgentProfile, nodes []v1.Node, profileAppliedPerNode map[string]types.NamespacedName,
+	now metav1.Time) (map[string]types.NamespacedName, error) {
+	conflicts := false
+	nodesThatMatchProfile := map[string]bool{}
+	profileStatus := datadoghqv1alpha1.DatadogAgentProfileStatus{}
+
+	if hash, err := comparison.GenerateMD5ForSpec(profile.Spec); err != nil {
+		logger.Error(err, "couldn't generate hash for profile", "name", profile.Name, "namespace", profile.Namespace)
+	} else {
+		profileStatus.CurrentHash = hash
+	}
 
-			if matchesNode {
-				if existingProfile, found := profileAppliedPerNode[node.Name]; found {
-					// Conflict. This profile should not be applied.
-					conflicts = true
-					logger.Info("conflict with existing profile, skipping", "conflicting profile", profile.Namespace+"/"+profile.Name, "existing profile", existingProfile.String())
-					break
-				} else {
-					nodesThatMatchProfile[node.Name] = true
-				}
-			}
-		}
+	if err := datadoghqv1alpha1.ValidateDatadogAgentProfileSpec(&profile.Spec); err != nil {
+		logger.Error(err, "profile spec is invalid, skipping", "name", profile.Name, "namespace", profile.Namespace)
+		profileStatus.Conditions = SetDatadogAgentProfileCondition(profileStatus.Conditions, NewDatadogAgentProfileCondition(ValidConditionType, metav1.ConditionFalse, now, InvalidConditionReason, err.Error()))
+		profileStatus.Valid = metav1.ConditionFalse
+		UpdateProfileStatus(profile, profileStatus, now)
+		return profileAppliedPerNode, err
+	}
 
-		if conflicts {
-			continue
+	for _, node := range nodes {
+		matchesNode, err := profileMatchesNode(profile, node.Labels)
+		if err != nil {
+			profileStatus.Conditions = SetDatadogAgentProfileCondition(profileStatus.Conditions, NewDatadogAgentProfileCondition(ValidConditionType, metav1.ConditionFalse, now, InvalidConditionReason, err.Error()))
+			profileStatus.Valid = metav1.ConditionFalse
+			UpdateProfileStatus(profile, profileStatus, now)
+			return profileAppliedPerNode, err
 		}
 
-		for node := range nodesThatMatchProfile {
-			profileAppliedPerNode[node] = types.NamespacedName{
-				Namespace: profile.Namespace,
-				Name:      profile.Name,
+		if matchesNode {
+			if existingProfile, found := profileAppliedPerNode[node.Name]; found {
+				// Conflict. This profile should not be applied.
+				conflicts = true
+				logger.Info("conflict with existing profile, skipping", "conflicting profile", profile.Namespace+"/"+profile.Name, "existing profile", existingProfile.String())
+				break
+			} else {
+				nodesThatMatchProfile[node.Name] = true
```

Comment:
I tested this out and it wouldn't work for profiles that conflict in like [this test case](https://github.com/DataDog/datadog-operator/pull/1155/files#diff-e68bb010d2321095701b73743b9c27b3b13a7edad2aa58193f9e2025ae701f93R267) (`several conflicting profiles with different creation timestamps`). In this test, we expect profile 1 to be discarded because it conflicts with profile 2, but setting `profileAppliedPerNode` here means profile 1 will be applied since we don't go through all the nodes to know if a profile conflicts